### ISSUE TITLE
only support batch_size=1 when I set indices_tuple in TripletMarginLoss?

### ISSUE THREAD

**Issue by DogeWatch** [KevinMusgrave/pytorch-metric-learning#135]
In my dataset.__getitem__, one index I return anchor, positve, negative and their labels, so in one minibatch, the labels' shape could be (batch_size, 3), and let indices_tuple=labels their was a wrong message because indices_tuple's length can only be 3 or 4. how can I specify the indices_tuple?

**Comment by KevinMusgrave** [comment]
The embeddings should have shape (batch_size, D), where D is the embedding size, and labels should have shape (batch_size). So in your example, your batch needs to be reshaped so that batch_size = 3. For example:
```python
# your current labels look something like this
labels = torch.LongTensor([[0, 0, 1]])

# they should look like this instead
labels = torch.LongTensor([0, 0, 1])
```

To use pre-packaged triplets like you're doing, you can either create ```indices_tuple```, or use ```EmbeddingsAlreadyPackagedAsTriplets```. See this previous issue for an example: https://github.com/KevinMusgrave/pytorch-metric-learning/issues/30#issuecomment-607465761

**Comment by DogeWatch** [comment]
thx, I've solved this problem. Is there a general way of torch.Dataset.__getitem__ to use all methods in the metric-learning? like randomly get two or three items per class?

**Comment by KevinMusgrave** [comment]
If you pass in a randomly sampled batch into one of the loss functions, and leave indices_tuple as None, then the loss functions will use all possible pairs/triplets in the batch. 

If you're talking getting samples in a class-balanced way, you can also look at [MPerClassSampler](https://kevinmusgrave.github.io/pytorch-metric-learning/samplers/#mperclasssampler), and pass the sampler into your dataloader.

**Comment by KevinMusgrave** [comment]
Closing for now, please re-open if you have more questions.
